azurelinuxagent/common/protocol/imds.py

```diff
@@ -254,38 +259,71 @@ def _http_get(self, endpoint, resource_path, headers):
+                                               resource_path, restutil.read_response_error(resp))
+
+        if restutil.request_failed(resp):
+            return IMDS_RESPONSE_ERROR, "IMDS error in /metadata/{0}: {1}".format(
+                                        resource_path, restutil.read_response_error(resp))
+
+        return IMDS_RESPONSE_SUCCESS, resp.read()
+
     def get_metadata(self, resource_path, is_health):
         """
         Get metadata from IMDS, falling back to Wireserver endpoint if necessary.
 
         :param str resource_path: path of IMDS resource
         :param bool is_health: True if for health/heartbeat, False otherwise
-        :return: Tuple<is_request_success:bool, response:str>
-            is_request_success: True when connection succeeds, False otherwise
-            response: response from IMDS on request success, failure message otherwise
+        :return: Tuple<is_request_success:bool, is_service_error:bool, response:str>
+            is_request_success: True for successful response, False otherwise
+            is_service_error: True when service returned an error, False for connection errors
+            response: response from IMDS on successful response, failure message otherwise
         """
         headers = self._health_headers if is_health else self._headers
         endpoint = IMDS_ENDPOINT
-        try:
-            resp = self._http_get(endpoint=endpoint, resource_path=resource_path, headers=headers)
-        except HttpError as e:
+
+        status, resp = self._get_metadata_from_endpoint(endpoint, resource_path, headers)
+        if status == IMDS_CONNECTION_ERROR:
             logger.periodic_warn(logger.EVERY_FIFTEEN_MINUTES,
                                  "[PERIODIC] Unable to connect to primary IMDS endpoint {0}".format(endpoint))
-            if not self._regex_imds_ioerror.match(str(e)):
-                raise
             endpoint = self._protocol_util.get_wireserver_endpoint()
-            try:
-                resp = self._http_get(endpoint=endpoint, resource_path=resource_path, headers=headers)
-            except HttpError as e:
-                logger.periodic_warn(logger.EVERY_FIFTEEN_MINUTES,
-                                     "[PERIODIC] Unable to connect to backup IMDS endpoint {0}".format(endpoint))
-                if not self._regex_imds_ioerror.match(str(e)):
-                    raise
-                return False, "IMDS error in /metadata/{0}: Unable to connect to endpoint".format(resource_path)
-        if restutil.request_failed(resp):
-            return False, "IMDS error in /metadata/{0}: {1}".format(
-                resource_path, restutil.read_response_error(resp))
-        return True, resp.read()
+            status, resp = self._get_metadata_from_endpoint(endpoint, resource_path, headers)
+
+        if status == IMDS_RESPONSE_SUCCESS:
+            return True, False, resp
```

Comment:
You could use NamedTuple to simplify the logic here.

Reply:
Great idea!